### ISSUE TITLE
How about we write articles about different Data Structures as well? 

### ISSUE THREAD

**Issue by ds-249** [FreeCodeCamp/wiki#1044]

**Comment by alayek** [comment]
Notwithstanding the fact that this would be a massive undertaking in itself, we have to use some statically typed, object oriented language for the code examples.

Our Wiki articles, at present, are mostly on JavaScript. I don't know of many material that would discuss even basic Stacks or Queues in JS.

We are, however, planning to add languages like Python or Java in our curriculum, and it is expected there would be some basic data structure problems in that. Which means we would also need corresponding articles in our Wiki on some basic data structures.

FreecodeCamp is about getting a candidate job ready, or improve on his current technical skills to move to a more technical role within their current day job. There are very few jobs that require complex Data Structure knowledge on a day-to-day basis. So it was never a priority for us.

However, interviews are a step to get a job, and the tech interviews tend to ask some tricky data structure problems - doubly linked queues, linked lists, trie etc.

It would massively help if you would like to take up this initiative and get the ball rolling. You can start writing in any standard language of your choice - Java or Python. Since we don't have any current plans for C or C++ in-browser challenges, we would rather not have articles in those.

**Comment by ds-249** [comment]
I code in python2, if that's okay, I can start right up. :)

**Comment by alayek** [comment]
@ds-249 that's cool! We are currently planning to support only Python 3 for now, so you raise the PRs in Python 2, and we would recommend changes and tweaks for Python 3.

You can discuss this more with us in our [chatroom](https://gitter.im/FreeCodeCamp/Wiki)

**Comment by abhisekp** [comment]
This might be helpful for this initiative :point_right: http://rosettacode.org/wiki/Category:Programming_Tasks

**Comment by alayek** [comment]
Udacity has recently launched a course titled "Technical Interviews" which covers some of these topics in Python.

**Comment by varunu28** [comment]
@alayek  I code in Python 3 and would like to get started on this. We can also add up few algorithms which get frequently asked in interviews.

**Comment by alayek** [comment]
@varunu28 for algorithms, I am under the impression you are talking about the sorting ones, median finding, string matching etc. (I am not including the ones that would be associated with a DS). Is that the case?

We have some Algorithm articles, but those are more for our internal in-browser challenges, and less for what an Algo course in college would cover. Sure, we could absolutely do that!

I think we should keep that as a separate epic though.

**Comment by varunu28** [comment]
@alayek Yeah I am talking about those and also like the ones for which you mentioned Udacity course.
Still we can keep this for future. I would get started on this.

**Comment by alayek** [comment]
@varunu28 please join our [chatroom](https://gitter.im/FreeCodeCamp/Wiki) to discuss this in detail.

**Comment by varunu28** [comment]
@alayek @ds-249 
Can we have a basic mock-up design of what the articles need to cover and what should be the format for it?

**Comment by ds-249** [comment]
@varunu28  I am following something like this

# Hash Tables and Hashing Functions

## Introduction to hashing

Hashing is designed to solve the problem of needing to efficiently find or store an item in a collection. 
For example, if we have a list of 10,000 words of English and we want to check if a given word is in the list, it would be inefficient to successively compare the word with all 10,000 items until we find a match. 
Hashing is a technique to make things more efficient by effectively narrowing down the search at the outset.

## What is hashing?

**TODO**

## How hashing works

**TODO**

**Comment by varunu28** [comment]
@ds-249 What exactly are we looking to put in "What is" and "How" part?
Will it be just the theory or also the implementation?

**Comment by ds-249** [comment]
@varunu28  I'll add another heading for Implementation in programs, where I'll add code snippets of python and java in it.

**Comment by varunu28** [comment]
ok. Do we need to add in both in Python and Java?

**Comment by varunu28** [comment]
I am starting with Arrays for this

**Comment by alayek** [comment]
@varunu28 ideally we would want a separate implementation section, and subsection with Python and another subsection with Java.

**Comment by Rafase282** [comment]
Adding a label to categorize these will be helpful @alayek particularly for when we move all articles into sub directories for organization in the near future hopefully.

**Comment by jainaman224** [comment]
# Linked list

 Linked list is a linear collection of data elements, called nodes pointing to the next node by means of a pointer. It is a data structure consisting of a group of nodes which together represent a sequence. It contains data and links to the next node (sometimes to previous nodes too) in the sequence.

**Types:**

1. Singly Linked List

    Singly linked lists contain nodes which have a `data` field as well as a `next` field, which points to the next node in the sequence. Operations that can be performed on singly linked lists are insertion, deletion and traversal.
    ```
     Singly Link List
    --------------
    
        head
         |
         |
       +-----+--+      +-----+--+      +-----+------+
       |  1  |o----->  |  2  |o----->  |  3  | NULL |
       +-----+--+      +-----+--+      +-----+------+

    ```

2. Doubly Linked List

    Doubly linked lists contain node which have `data` field, `next` field and another link field `prev` pointing to the previous node in the sequence.

    ```
    Doubly Linked List
    ----------------

               head
                |
                |
       +------+-----+--+    +--+-----+--+       +-----+------+
       |      |     |o------>  |     |o------>  |     |      |
       | NULL |  1  |          |  2  |          |  3  | NULL |
       |      |     |  <------o|     |  <------o|     |      |
       +------+-----+--+    +--+-----+--+       +-----+------+

    ```

3. Circular Linked List

    Circular linked lists is a singly linked list in which last node, `next` field points to first node in the sequence.

    ```
    Circular Linked List
     ------------------

           head
            |
            |
         +-----+--+      +-----+--+      +-----+--+
  -->  |  1  |o----->  |  2  |o----->  |  3  |o----
  |    +-----+--+      +-----+--+      +-----+--+  |
  |                                                |
   ------------------------------------------------
    ```

4. Circular Doubly Linked list

    Circular linked lists is a doubly linked list in which last node, `next` points to first node of the sequence and first node, `prev` points to last node of the sequence.

**Implementation:**

Python Implementation of Singly Linked List

```python

class Node(object):
    # Constructor
    def __init__(self, data=None, next=None):
        self.data = data
        self.next = next

    # Function to get data
    def get_data(self):
        return self.data

    # Function to get next node
    def get_next(self):
        return self.next

    # Function to set next field
    def set_next(self, new_next):
        self.next = new_next


class LinkedList(object):
    def __init__(self, head=None):
        self.head = head

    # Function to insert data
    def insert(self, data):
        # new_node is a object of class Node
        new_node = Node(data)
        new_node.set_next(self.head)
        self.head = new_node
        print("Node with data " + str(data) + " is created succesfully")

    # Function to get size
    def size(self):
        current = self.head
        count = 0
        while current:
            count += 1
            current = current.get_next()
        print("Size of link list is " + str(count))

    # Function to search a data
    def search(self, data):
        current = self.head
        found = False
        while current and found is False:
            if current.get_data() == data:
                found = True
            else:
                current = current.get_next()
        if current is None:
            print("Node with data " + str(data) + " is not present")
        else:
        	print("Node with data " + str(data) + " is found")


    # Function to delete a node with data
    def delete(self, data):
        current = self.head
        previous = None
        found = False
        while current and found is False:
            if current.get_data() == data:
                found = True
            else:
                previous = current
                current = current.get_next()
        if current is None:
            print("Node with data " + str(data) + " is not in list")
        elif previous is None:
            self.head = current.get_next()
            print("Node with data " + str(data) + " is deleted successfully")
        else:
            previous.set_next(current.get_next())
            print("Node with data " + str(data) + " is deleted successfully")

SLL = LinkedList() # Creates an object of class LinkedList
SLL.size() # prints 'Size of link list is 0'
data_elements = [5, 10, 2, 6, 8, 20]
for each in data_elements:
	SLL.insert(each)
SLL.size() # prints 'Size of link list is 6'
SLL.search(4) # prints 'Node with data 4 is not present'
SLL.search(5) # prints 'Node with data 5 is found'
SLL.delete(4) # prints 'Node with data 4 is not in list'
SLL.delete(5) # prints 'Node with data 5 is deleted successfully'
```
:rocket: [Run Code](https://repl.it/CVq3)


**Advantages**

1. Linked lists are a dynamic data structure, which can grow and shrink, allocating and deallocating memory while the program is running.
2. Insertion and deletion of node are easily implemented in a linked list at any position.

**Disadvantages**

1. They use more memory than arrays because of the memory used by their pointers (`next` and `prev`).
2. Random access is not possible in linked list. We have to access nodes sequentially.

**Comment by varunu28** [comment]
@alayek  I want to get started with stacks. Please assign that to me.

**Comment by ilogin** [comment]
Please assign Trees - Binary Search Trees to me.
